File: MyApp.Api/WidgetController.cs
using MyApp.Data;
using System.Data.Entity.Infrastructure;
using System.Threading;
using System.Threading.Tasks;
using System.Web.Http;

namespace MyApp.Api
{
    [RoutePrefix("api/widgets")]
    public class WidgetController : BaseController<MyDbContext>
    {
        public WidgetController(IDbContextFactory<MyDbContext> dbContextFactory) : base(dbContextFactory) { }        

        [HttpGet, Route("{id}")]
        public Task<WidgetDto> Get(CancellationToken cancellationToken, int id)
        {
            return GetSingleOrDefault<Widget, WidgetDto>(cancellationToken, w => w.Id == id);
        }

        [HttpGet, Route]
        public Task<WidgetDto[]> GetAll(CancellationToken cancellationToken)
        {
            return GetArray<Widget, WidgetDto>(cancellationToken);
        }

        [HttpPost, Route]
        public Task<int> Insert(CancellationToken cancellationToken, WidgetDto dto)
        {
            return Insert<Widget>(cancellationToken, dto);
        }

        [HttpPut, Route("{id}")]
        public Task Update(CancellationToken cancellationToken, int id, WidgetDto dto)
        {
            return Update<Widget>(cancellationToken, id, dto);
        }

        [HttpDelete, Route("{id}")]
        public Task Delete(CancellationToken cancellationToken, int id)
        {
            return Delete<Widget>(cancellationToken, id);
        }
    }
}

File: MyApp.WebApplication/Typewriter/WidgetController.ts
class WidgetService {
    constructor(private $http: ng.IHttpService) { }
    get(id: number): ng.IHttpPromise<Widget> {            
        return this.$http<Widget>({
            url: `api/widgets/${id}`, 
            method: "get", 
            data: typeof(null) === 'string' ? '"' + null + '"' : null
        });
    }
    getAll(): ng.IHttpPromise<Widget[]> {            
        return this.$http<Widget[]>({
            url: `api/widgets`, 
            method: "get", 
            data: typeof(null) === 'string' ? '"' + null + '"' : null
        });
    }
    insert(dto: Widget): ng.IHttpPromise<number> {            
        return this.$http<number>({
            url: `api/widgets`, 
            method: "post", 
            data: typeof(dto) === 'string' ? '"' + dto + '"' : dto
        });
    }
    update(id: number, dto: Widget): ng.IHttpPromise<void> {            
        return this.$http<void>({
            url: `api/widgets/${id}`, 
            method: "put", 
            data: typeof(dto) === 'string' ? '"' + dto + '"' : dto
        });
    }
    delete(id: number): ng.IHttpPromise<void> {            
        return this.$http<void>({
            url: `api/widgets/${id}`, 
            method: "delete", 
            data: typeof(null) === 'string' ? '"' + null + '"' : null
        });
    } 
} 
angular.module("app").service("WidgetService", ["$http", WidgetService]);
